src/test/fs_tests.cpp

```diff
@@ -11,6 +11,26 @@
 
 BOOST_FIXTURE_TEST_SUITE(fs_tests, BasicTestingSetup)
 
+BOOST_AUTO_TEST_CASE(fsbridge_pathtostring)
+{
+    std::string u8_str = "fs_tests_₿_🏃";
+    BOOST_CHECK_EQUAL(fs::PathToString(fs::PathFromString(u8_str)), u8_str);
+    BOOST_CHECK_EQUAL(fs::u8path(u8_str).u8string(), u8_str);
+    BOOST_CHECK_EQUAL(fs::PathFromString(u8_str).u8string(), u8_str);
+    BOOST_CHECK_EQUAL(fs::PathToString(fs::u8path(u8_str)), u8_str);
+#ifndef WIN32
+    // On non-windows systems, verify that non-UTF8 strings can be round
+    // tripped successfully. On windows, file paths are natively represented as
+    // unicode strings, so an invalid UTF-8 sequence that does not map to any
```

Comment:
nit: I find the sentence "On windows, file paths are natively represented as unicode strings, [..]" slightly confusing because of that "unicode strings". My immediate understanding is something like "a sequence of bytes in either UTF-8, UTF-16, etc. encoding" which makes me guess for no good reason :)

According to https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160#syntax (and your [comment](https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-928112127) too), it holds that *Windows uses a null-terminated sequence of wchar_t, encoded as UTF-16 (one or more elements for each character).*

So I would expect something like "On Windows, file paths are natively represented by wchar_t sequences, encoded as UTF-16 strings, [..]" 

Hopefully, I get it right.

Reply:
> nit: I find the sentence "On windows, file paths are natively represented as unicode strings, [..]" slightly confusing because of that "unicode strings". 

Thanks rewrote this and replaced "unicode strings" with "unicode". The character types and encodings that windows uses internally are not really relevant here. The point is just that paths on windows are unicode and have to be encoded as bytes, while paths on posix are already bytes and do not get encoded.

> My immediate understanding is something like "a sequence of bytes in either UTF-8, UTF-16, etc. encoding" which makes me guess for no good reason :)

This isn't actually true, see https://github.com/rust-lang/rust/issues/12056 as a starting point if you are really curious about this. But this comment is just explaining why the test is skipped on windows.

> According to https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160#syntax (and your [comment](https://github.com/bitcoin/bitcoin/pull/22937#issuecomment-928112127) too), it holds that _Windows uses a null-terminated sequence of wchar_t, encoded as UTF-16 (one or more elements for each character)._

This seems to be a simplification, they are wchar_t, but not strictly valid utf-16. Not too relevant to the API in this PR, though, which is just trying to preserve current behavior and use a safe string encoding.